### PR TITLE
✨ Configuration assembly cumulative overrides

### PR DESCRIPTION
**Pull request by muchzill4**
### Description

Added logic for assembly of cumulative config attributes.

- I've split the modifiers into separate modules, as the file was growing a bit (the code for `SetConfigValue` remains the same
- For reasons `pytest` stopped picking up `factories`, so I had to add all the `__init__.py` files

Integration test of what functionality we're after is probably the best place to start: https://github.com/ARMmbed/mbed-build/pull/20/files#diff-54c67391e6009657b93988e42bb40fc7

Then, where the modifier is built from entry in `"target_overrides": {}` JSON: https://github.com/ARMmbed/mbed-build/pull/20/files#diff-a47ca7fe3cbe3f7c8e12eb8cc70f2f6bR16

---

FWIW, this does not cover `BOOTLOADER_OVERRIDES` and doesn't really handle `target.<foo>` style regular config overrides well - for those `target.` prefix will need to be stripped.
 
I don't want to complicate this PR more than it needs to be right now. The only checkbox we're trying to tick here is cumulative overrides.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_build/_internal/config/config_modifiers/cumulative.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        override_key = cast(CumulativeOverrideKey, override_key)
+        if override_type == "add":
+            return AppendToConfig(key=override_key, value=data)
+        if override_type == "remove":
+            return RemoveFromConfig(key=override_key, value=data)
+
+    raise UnableToBuildCumulativeModifier
+
+
+CumulativeOverrideKey = Literal["features"]
```
Comment (muchzill4) [inline mbed_build/_internal/config/config_modifiers/cumulative.py:56]: Sadly, I wasn't able to figure out how to use pre-defined `ACCUMULATING_OVERRIDES` with `Literal[]`.

I've attempted both `typing.get_args(CumulativeOverrideKey)` and `Literal[list(ACCUMULATING_OVERRIDES)]`.

File: mbed_build/_internal/config/config_modifiers/cumulative.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+class UnableToBuildCumulativeModifier(Exception):
+    """Raised when data given to builder is not valid for cumulative modifier."""
+
+
+# TODO: potentially add more accumulating properties here.
+# Two projects this has been tested against:
+# - mbed-cloud-client-example
+# - mbed-os-example-blinky
+# only contain "features_add" and "features_remove".
+ACCUMULATING_OVERRIDES = ("features",)
```
Comment (rosiemiller2) [inline mbed_build/_internal/config/config_modifiers/cumulative.py:26]: There is the full list here: https://github.com/ARMmbed/mbed-targets/blob/master/mbed_targets/_internal/targets_json_parsers/accumulating_attribute_parser.py#L15

Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/cumulative.py:26]: Oh yeah, I had a full list before here, but decided to go only with what we need right now. My thought process was - let's not add things to config we're not 100% sure is necessary right now. I could link to that full list, thoughts?

I'm a bit concerned most of these are `targets.json` specific. 🤔

Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/cumulative.py]: I've added a full list of those attributes.

File: mbed_build/_internal/config/config_modifiers/utils.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Utilities for building config modifiers."""
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config_modifiers/utils.py:5]: Could have a more descriptive name for this file? `utils` is a bit ambiguous tbh. How about `build_modifiers`?

File: mbed_build/_internal/config/config_modifiers/set_config_setting.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    key: str
+    value: Any
+    help: Optional[str]
+
+    def __call__(self, config: "Config") -> None:
+        """Mutate config by overwriting existing entry with new data."""
+        existing = config["settings"].get(self.key)
+        if existing:
+            config["settings"][self.key]["value"] = self.value
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config_modifiers/set_config_setting.py:45]: Does `existing["value"]` not work here?

File: mbed_build/_internal/config/config_modifiers/utils.py (hunk 3 above)
Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/utils.py]: https://github.com/ARMmbed/mbed-build/pull/20/commits/f0b8788d13b0ade5e0c2637ca6ca1c8c1d541227

File: mbed_build/_internal/config/config_modifiers/set_config_setting.py (hunk 4 above)
Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/set_config_setting.py]: Welp. I've missed that bit when rebasing! Nice catch. https://github.com/ARMmbed/mbed-build/pull/20/commits/99e6645261868e819478dd1cdb62d2e9a7575629

File: mbed_build/_internal/config/config_modifiers/set_target_attribute.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+
+ACCUMULATING_OVERRIDES = ("extra_labels", "macros", "device_has", "features", "components")
+MODIFIERS = ("add", "remove")
+ALL_ACCUMULATING_OVERRIDES = ACCUMULATING_OVERRIDES + tuple(
+    f"{attr}_{suffix}" for attr, suffix in itertools.product(ACCUMULATING_OVERRIDES, MODIFIERS)
+)
+
+
+def build(key: str, data: Any) -> Callable:
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config_modifiers/set_target_attribute.py:29]: Could we have an `Args` section in the docstring describing what `data` is? The type is `Any` and the name is fairly generic so it's hard to understand what `data` actually is.

File: mbed_build/_internal/config/config_modifiers/set_config_setting.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    def __call__(self, config: "Config") -> None:
+        """Mutate config by overwriting existing entry with new data."""
+        existing = config["settings"].get(self.key)
+        if existing:
+            existing["value"] = self.value
+        else:
+            config["settings"][self.key] = {
+                "value": self.value,
+                "help": self.help,
+            }
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config_modifiers/set_config_setting.py:50]:
```suggestion
        existing = config["settings"].get(self.key, {})
        if existing:
            existing["value"] = self.value
        else:
            existing = {
                "value": self.value,
                "help": self.help,
            }
```

Thoughts?

Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/set_config_setting.py]: Awesome!

File: mbed_build/_internal/config/config_modifiers/set_target_attribute.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    A key for this modifier needs to follow a spec from old tools implementation:
+    - start with "target."
+    - the part after "target." is one of ACCUMULATING_OVERRIDES
+    """
+    # Key following a spec would:
+    if not key.startswith("target."):
+        raise InvalidModifierData
+
+    key = key[7:]  # Strip "target." prefix
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config_modifiers/set_target_attribute.py:40]:
```suggestion
    prefix = "target."
    if not key.startswith(prefix):
        raise InvalidModifierData

    key = key[len(prefix):]  # Strip "target." prefix
```

File: mbed_build/_internal/config/config_modifiers/set_config_setting.py (hunk 6 above)
Comment (muchzill4) [reply mbed_build/_internal/config/config_modifiers/set_config_setting.py]: Damn it, it doesn't really work. `existing` is not a reference to the actual dict entry.

Comment (rwalton-arm) [reply mbed_build/_internal/config/config_modifiers/set_config_setting.py:50]: Oh, right. I guess `get` is returning a new dictionary. Apologies I didn't actually test it.

**Pull request merged**